persistence-modules/spring-boot-persistence-h2/src/main/java/com/baeldung/h2db/lazy_load_no_trans/service/ServiceLayer.java

```diff
@@ -0,0 +1,36 @@
+package com.baeldung.h2db.lazy_load_no_trans.service;
+
+import com.baeldung.h2db.lazy_load_no_trans.entity.Document;
+import com.baeldung.h2db.lazy_load_no_trans.entity.User;
+import com.baeldung.h2db.lazy_load_no_trans.repository.UserRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Service
+public class ServiceLayer {
+
+    @Autowired
+    private UserRepository userRepository;
+
+    @Transactional(readOnly = true)
+    public int countAllDocsTransactional() {
+        return countAllDocs();
+    }
+
+    public int countAllDocsNonTransactional() {
+        return countAllDocs();
+    }
+
+    private int countAllDocs() {
+        List<Document> docs = userRepository
+            .findAll().stream()
+            .flatMap(user -> user.getDocs().stream())
+            .collect(Collectors.toList());
+        return docs.size();
```

Comment:
You don't need to collect to a list and then return the size of the list. You can go

```java
return userRepository.findAll()
    .stream()
    .map(User::getDocs)
    .flatMap(Document::stream)
    .count();
```

Or similar.

Reply:
That's right, fixed